contracts/core/liquidators/twap-impl/TWAPAuction.sol

```diff
@@ -39,4 +56,357 @@ contract TWAPAuction is TwoAssetPriceBoundedLinearAuction {
+     * @param _assetPairHash            Hash of asset pair being rebalanceds
+     * @param _rebalanceVolume          Value of collateral being rebalanced
+     */
+    function validateTWAPParameters(
+        TWAPLiquidatorData memory _twapLiquidatorData,
+        bytes32 _assetPairHash,
+        uint256 _rebalanceVolume
+    )
+        internal
+        view
+    {
+        require(
+            chunkSizeWhiteList[_assetPairHash].isWithin(_twapLiquidatorData.usdChunkSize),
+            "TWAPAuction.validateTWAPParameters: Passed chunk size must be between bounds."
+        );
+
+        uint256 chunkAuctions = _rebalanceVolume.divCeil(_twapLiquidatorData.usdChunkSize);
+        uint256 expectedTWAPAuctionTime = chunkAuctions.mul(expectedChunkAuctionLength)
+            .add(chunkAuctions.sub(1).mul(_twapLiquidatorData.chunkAuctionPeriod));
+
+        uint256 failAuctionPeriod = IRebalancingSetTokenV3(msg.sender).failAuctionPeriod();
+
+        require(
+            expectedTWAPAuctionTime < failAuctionPeriod,
+            "TWAPAuction.validateTWAPParameters: Expected auction duration exceeds allowed length."
+        );
+    }
+
+    /**
+     * Validates next chunk auction can begin.
+     *
+     * @param _twapAuction                  TWAPAuction State object
+     */
+    function validateNextChunkAuction(
+        TWAPState storage _twapAuction
+    )
+        internal
+        view
+    {
+        Auction.validateAuctionCompletion(_twapAuction.chunkAuction.auction);
+
+        require(
+            isRebalanceActive(_twapAuction),
+            "TWAPState.validateNextChunkAuction: TWAPAuction is finished."
+        );
+
+        require(
+            _twapAuction.lastChunkAuctionEnd.add(_twapAuction.chunkAuctionPeriod) <= block.timestamp,
+            "TWAPState.validateNextChunkAuction: Not enough time elapsed from last chunk auction end."
+        );
+    }
+
+    /**
+     * Checks if the amount of Sets still to be auctioned (in aggregate) is greater than the minimumBid
+     *
+     * @param _twapAuction                  TWAPAuction State object
+     */
+    function isRebalanceActive(
+        TWAPState storage _twapAuction
+    )
+        internal
+        view
+        returns (bool)
+    {
+        // Sum of remaining Sets in current chunk auction and order remaining
+        uint256 totalRemainingSets = _twapAuction.orderRemaining.add(
+            _twapAuction.chunkAuction.auction.remainingCurrentSets
+        );
+        
+        // Check that total remaining sets is greater than minimumBid
+        return totalRemainingSets > _twapAuction.chunkAuction.auction.minimumBid;
+    }
+
+    /**
+     * Calculates chunkSize of auction in current Set terms
+     *
+     * @param _totalSetAmount       Total amount of Sets being auctioned
+     * @param _rebalanceVolume      The total value being auctioned
+     * @param _usdChunkSize         Value of chunk size
+     */
+    function calculateChunkSize(
```

Comment:
I feel like our variables and functions need to do a better just describing its denomination, particularly where things can be denominated in USD, currentSets, etc. It can get confusing if not explicit.

Ideally, we have a typing for this. But if we don't, we can enforce it in the variable name.

Reply:
Disagree, it should be agnostic because in theory we could use any denomination we want. If anything I think we're actually being overly restrictive and should be NOT explicitly defining it.